### PR TITLE
Optimise cycling objects

### PR DESCRIPTION
**Pull request by oliver-sanders**
Efficiency changes which reduce the memory and CPU impact of cycling objects:
 - Avoid creating multiple parser instances.
 - Move regex compilation out of the `__init__` method.
 - Slot classes.
 - Minor fiddling with exclusions objects (@dvalters could you check these changes).

In combination with slotting of the isodatetime library (https://github.com/metomi/isodatetime/pull/74) this reduces the resource requirements of suites with multiple sequences defined in the graphing.

The following results were obtained running `cylc validate` over the following suite varying the value of `sequences`:

```
[scheduling]
    initial cycle point = 2000
    spawn to max active cycle points = true
    max active cycle points = 5
    [[dependencies]]
{% for seq_no in range(sequences | default(10) | int) %}
    {% set minutes = seq_no % 60 %}
    {% set hours = ((seq_no / 60) | int) % 60 %}
        [[[R{{cycles | default(1)}}/+PT{{hours}}H{{minutes}}M]]]
            graph = task
{% endfor %}
```

memory
cpu-time

I'll include this suite in the standard profile-battery tests in a future pull.
The impact on the runtime of the same suite:

memory
cpu-time

The complex suite shows a small reduction in memory.

### CODE REVIEW

**Comment by hjoliver** [comment]
Nice result.  Definitely include your new test suite in the profile-battery.

File: lib/cylc/cycling/iso8601.py [hunk 1]
```diff
@@ -288,9 +293,12 @@ class ISO8601Exclusions(ExclusionBase):
     grouped exclusion sequences. The Python ``in`` and ``not in`` operators
     may be used on this object to determine if a point is in the collection
     of exclusion sequences."""
+
+    __slots__ = ExclusionBase.__slots__ + ('p_iso_exclusions',)
+
     def __init__(self, excl_points, start_point, end_point=None):
         super(ISO8601Exclusions, self).__init__(start_point, end_point)
-        self.p_iso_exclusions = set()
+        self.p_iso_exclusions = []
```
Comment (hjoliver) [inline lib/cylc/cycling/iso8601.py]: What's the rationale behind this change?  I would have thought `set` was better, with the need for membership testing below...

Comment (hjoliver) [inline lib/cylc/cycling/iso8601.py]: (this comment was meant to apply to `self.exclusion_points`)

Comment (hjoliver) [inline lib/cylc/cycling/iso8601.py]: OK, from a bit of googling, I guess `set` is implemented rather like `dict`, hence much higher memory consumption than `list`.

Comment (hjoliver) [inline lib/cylc/cycling/iso8601.py]: (Answered my own question, will merge ...)

Comment (oliver-sanders) [inline lib/cylc/cycling/iso8601.py]: The reason I changed this is because an empty set is much larger than an empty list:

Size in bytes for collections of `None` elements:

elements | list | set
--- | --- | ---
0 | 72 | 232
1 | 96 | 248
2 | 104 | 248
3 | 112 | 248
4 | 120 | 248
5 | 128 | 248
6 | 136 | 248
7 | 144 | 248
8 | 152 | 248
9 | 160 | 248

Comment (dvalters) [inline lib/cylc/cycling/iso8601.py]: This suite doesn't test effect on performance when exclusions are used, unless I'm missing something? (Although it tests 'empty'/no exclusions, which I suppose would be majority of the use cases?)

Comment (oliver-sanders) [inline lib/cylc/cycling/iso8601.py]: > no exclusions, which I suppose would be majority of the use cases

The vast majority of sequence objects will not have exclusions. In these cases we have `sequence.exclusions = []` so the memory impact is minimal and `p_iso_exclusions` doesn't even come into it.

The reason for changing to a set was to make exclusion objects a little slimmer - it's pretty marginal I'll admit but many of these changes are, a few bytes saved here, a few there.
